File: src/models/compression.test.ts
import { describe, it, expect } from 'vitest'
import { build_merge_cache, get_compressed } from './compression'
import type { Slice } from './types'

function makeSlice(ts: number, dur: number, state: string, name: string | null = null): Slice {
  return { ts, dur, name, state, depth: null, io_wait: null, blocked_function: null }
}

describe('build_merge_cache', () => {
  it('returns empty map for empty input', () => {
    const cache = build_merge_cache([])
    expect(cache.size).toBe(0)
  })

  it('caches the original length', () => {
    const slices = [makeSlice(0, 100, 'Running'), makeSlice(100, 200, 'Sleeping')]
    const cache = build_merge_cache(slices)
    expect(cache.has(2)).toBe(true)
    expect(cache.get(2)!.length).toBe(2)
  })

  it('produces levels down to 2 for multi-slice input', () => {
    const slices = [
      makeSlice(0, 100, 'Running'),
      makeSlice(100, 200, 'Sleeping'),
      makeSlice(300, 150, 'Running'),
      makeSlice(450, 50, 'Runnable'),
    ]
    const cache = build_merge_cache(slices)
    expect(cache.has(4)).toBe(true)
    expect(cache.has(2)).toBe(true)
  })

  it('preserves tsRel relative to first slice', () => {
    const slices = [makeSlice(1000, 100, 'Running'), makeSlice(1100, 200, 'Sleeping')]
    const cache = build_merge_cache(slices)
    const full = cache.get(2)!
    expect(full[0].tsRel).toBe(0)
    expect(full[1].tsRel).toBe(100)
  })

  it('merging identical states costs less than different states', () => {
    const same = [
      makeSlice(0, 100, 'Running'),
      makeSlice(100, 100, 'Running'),
      makeSlice(200, 100, 'Sleeping'),
    ]
    const cache = build_merge_cache(same)
    // At level 2, the two Running slices should merge first (lower cost)
    const merged = cache.get(2)!
    expect(merged.length).toBe(2)
    expect(merged[0].state).toBe('Running')
    expect(merged[0].dur).toBe(200)
  })
})

describe('get_compressed', () => {
  const slices = [
    makeSlice(0, 100, 'Running'),
    makeSlice(100, 200, 'Sleeping'),
    makeSlice(300, 150, 'Running'),
    makeSlice(450, 50, 'Runnable'),
  ]
  const cache = build_merge_cache(slices)

  it('returns full sequence when target >= origN', () => {
    const result = get_compressed(cache, 4, 4)
    expect(result.length).toBe(4)
  })

  it('returns full sequence when target > origN', () => {
    const result = get_compressed(cache, 4, 100)
    expect(result.length).toBe(4)
  })

  it('clamps target below 2 to 2', () => {
    const result = get_compressed(cache, 4, 0)
    expect(result.length).toBe(2)
  })

  it('returns closest level >= target', () => {
    const result = get_compressed(cache, 4, 3)
    expect(result.length).toBe(3)
  })
})

File: android/app/src/test/java/com/swiperf/app/data/compression/CompressionTest.kt
package com.swiperf.app.data.compression

import com.swiperf.app.data.model.Slice
import org.junit.Assert.*
import org.junit.Test

class CompressionTest {

    private fun makeSlice(ts: Long, dur: Long, state: String? = "Running", name: String? = null): Slice {
        return Slice(ts = ts, dur = dur, name = name, state = state, depth = null, ioWait = null, blockedFunction = null)
    }

    @Test
    fun buildMergeCache_emptyInput_returnsEmptyCache() {
        val cache = Compression.buildMergeCache(emptyList())
        assertTrue(cache.isEmpty())
    }

    @Test
    fun buildMergeCache_containsOriginalLength() {
        val slices = listOf(
            makeSlice(0, 100, "Running"),
            makeSlice(100, 200, "Sleeping"),
            makeSlice(300, 150, "Running"),
            makeSlice(450, 50, "Sleeping"),
        )
        val cache = Compression.buildMergeCache(slices)
        assertTrue(cache.containsKey(4))
        assertEquals(4, cache[4]!!.size)
    }

    @Test
    fun buildMergeCache_cacheHasDecreasingLevels() {
        val slices = (0 until 5).map { makeSlice(it.toLong() * 100, 100, if (it % 2 == 0) "Running" else "Sleeping") }
        val cache = Compression.buildMergeCache(slices)
        assertTrue(cache.containsKey(5))
        assertTrue(cache.containsKey(2))
        for ((k, v) in cache) assertEquals(k, v.size)
    }

    @Test
    fun buildMergeCache_preservesTsRel() {
        val slices = listOf(
            makeSlice(1000, 100, "Running"),
            makeSlice(1100, 200, "Sleeping"),
        )
        val cache = Compression.buildMergeCache(slices)
        val seq = cache[2]!!
        assertEquals(0L, seq[0].tsRel)
        assertEquals(100L, seq[1].tsRel)
    }

    @Test
    fun buildMergeCache_identicalStatesAreCheaper() {
        val slices = listOf(
            makeSlice(0, 100, "Running"),
            makeSlice(100, 100, "Running"),
            makeSlice(200, 100, "Sleeping"),
        )
        val cache = Compression.buildMergeCache(slices)
        val merged2 = cache[2]!!
        // Same-state slices should merge first, so one of the two-element results has a merged Running
        val totalDur = merged2.sumOf { it.dur }
        assertEquals(300L, totalDur)
    }

    @Test
    fun getCompressed_returnsCorrectLevel() {
        val slices = (0 until 10).map { makeSlice(it.toLong() * 100, 100, "Running") }
        val cache = Compression.buildMergeCache(slices)
        val seq = Compression.getCompressed(cache, 10, 5)
        assertTrue(seq.size >= 5)
        assertTrue(seq.size <= 10)
    }

    @Test
    fun getCompressed_targetBelowTwo_clampedToTwo() {
        val slices = listOf(makeSlice(0, 100, "Running"), makeSlice(100, 100, "Sleeping"))
        val cache = Compression.buildMergeCache(slices)
        val seq = Compression.getCompressed(cache, 2, 1)
        assertEquals(2, seq.size)
    }
}
